### PR TITLE
[API/OP]Add a new API paddle.diagonal

### PR DESCRIPTION
**Pull request by zhangbo9674**
### PR types
New features

### PR changes
APIs

### Describe
This OP computes the diagonals of the input tensor x.

If x is 2D, returns the diagonal.
If x has larger dimensions, diagonals be taken from the 2D planes specified by axis1 and axis2. 
By default, the 2D planes formed by the first and second axis of the input tensor x.

The argument offset determines where diagonals are taken from input tensor x:
    - If offset = 0, it is the main diagonal.
    - If offset > 0, it is above the main diagonal.
    - If offset < 0, it is below the main diagonal.
    
Args:
```
x(Tensor): The input tensor x. Must be at least 2-dimensional. The input data type should be int32, int64, float32, float64.
offset(int, optional): Which diagonals in input tensor x will be taken. Default: 0 (main diagonals).
axis1(int, optional): The first axis with respect to take diagonal. Default: 0.
axis2(int, optional): The second axis with respect to take diagonal. Default: 1.
name (str, optional): Normally there is no need for user to set this property. For more information, please refer to :ref:`api_guide_Name`. Default: None.
```
if attr `axis1` or `axis2` is over the dims if imput Tensor, it will return a OutOfRange error.
if attr `offset` is over the range of [axis1, axis2], it will set the new dim as 0.

Returns:
        Tensor: the output data type is the same as input data type.

Examples:
```
import paddle

x = paddle.rand([2,2,3],'float32')
print(x)
# Tensor(shape=[2, 2, 3], dtype=float32, place=CUDAPlace(0), stop_gradient=True,
#        [[[0.45661032, 0.03751532, 0.90191704],
#          [0.43760979, 0.86177313, 0.65221709]],

#         [[0.17020577, 0.00259554, 0.28954273],
#          [0.51795638, 0.27325270, 0.18117726]]])

out1 = paddle.diagonal(x)
print(out1)
#Tensor(shape=[3, 2], dtype=float32, place=CUDAPlace(0), stop_gradient=True,
#       [[0.45661032, 0.51795638],
#        [0.03751532, 0.27325270],
#        [0.90191704, 0.18117726]])

out2 = paddle.diagonal(x, offset=0, axis1=2, axis2=1)
print(out2)
#Tensor(shape=[2, 2], dtype=float32, place=CUDAPlace(0), stop_gradient=True,
#       [[0.45661032, 0.86177313],
#        [0.17020577, 0.27325270]])

out3 = paddle.diagonal(x, offset=1, axis1=0, axis2=1)
print(out3)
#Tensor(shape=[3, 1], dtype=float32, place=CUDAPlace(0), stop_gradient=True,
#       [[0.43760979],
#        [0.86177313],
#        [0.65221709]])

out4 = paddle.diagonal(x, offset=0, axis1=1, axis2=2)
print(out4)
#Tensor(shape=[2, 2], dtype=float32, place=CUDAPlace(0), stop_gradient=True,
#       [[0.45661032, 0.86177313],
#        [0.17020577, 0.27325270]])
```
API doc：
图片
图片

### CODE REVIEW

File: paddle/fluid/operators/diagonal_op.cc [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+
+namespace paddle {
+namespace operators {
+
+class DiagonalOp : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+  void InferShape(framework::InferShapeContext *ctx) const override {
+    PADDLE_ENFORCE_EQ(
```
Comment (zhiqiu) [inline paddle/fluid/operators/diagonal_op.cc]: Better use `OP_INOUT_CHECK`

File: paddle/fluid/operators/diagonal_op.cc [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+        ctx->HasInput("Input"), true,
+        platform::errors::NotFound("Input of DiagonalOp is not found."));
+
+    PADDLE_ENFORCE_EQ(
+        ctx->HasOutput("Out"), true,
+        platform::errors::NotFound("Output of DiagonalOp is not found."));
+
+    int offset_ = ctx->Attrs().Get<int>("offset");
+    int dim1 = ctx->Attrs().Get<int>("axis1");
```
Comment (zhiqiu) [inline paddle/fluid/operators/diagonal_op.cc]: Maybe `dim1`->`axis1` is better, to be consistent with op's attr.

File: paddle/fluid/operators/diagonal_op.cc [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+    PADDLE_ENFORCE_LT(
+        dim2_, x_dims.size(),
+        platform::errors::OutOfRange(
+            "Attr(axis2) is out of range (expected to be in range of [%ld, "
+            "%ld], but got %ld).",
+            -(x_dims.size()), (x_dims.size() - 1), dim2));
+    PADDLE_ENFORCE_NE(dim1_, dim2_,
+                      platform::errors::InvalidArgument(
+                          "The dimensions should not be identical "
+                          "%ld vs %ld.",
```
Comment (zhiqiu) [inline paddle/fluid/operators/diagonal_op.cc]: %d for int

File: paddle/fluid/operators/diagonal_op.cc [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+                          dim1, dim2));
+
+    auto out_dims = vectorize(x_dims);
+    auto dim1_size = out_dims[dim1_];
+    auto dim2_size = out_dims[dim2_];
+    out_dims.erase(out_dims.begin() + std::max(dim1_, dim2_));
+    out_dims.erase(out_dims.begin() + std::min(dim1_, dim2_));
```
Comment (zhiqiu) [inline paddle/fluid/operators/diagonal_op.cc]: add some comments

File: paddle/fluid/operators/diagonal_op.cc [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+ public:
+  void Make() override {
+    AddInput("Input",
+             "(Tensor) The input tensor, from which the diagonals are taken.");
+    AddOutput(
+        "Out",
+        "(Tensor) The partial view of input with the its diagonal elements.");
+    AddAttr<int>(
+        "offset",
+        R"DOC((int, default 0), offset of the diagonal from the main diagonal. Can be both positive and negative. Defaults: 0.
```
Comment (zhiqiu) [inline paddle/fluid/operators/diagonal_op.cc]: Defaults -> Default

File: paddle/fluid/operators/diagonal_op.cu [hunk 6]
```diff
@@ -0,0 +1,362 @@
+/* Copyright (c) 2019 PaddlePaddle Authors. All Rights Reserved.
```
Comment (zhiqiu) [inline paddle/fluid/operators/diagonal_op.cu]: 2019 -> 2021

File: paddle/fluid/operators/diagonal_op.cu [hunk 7]
```diff
@@ -0,0 +1,362 @@
...
+        temp_stride = temp_stride * input_dim[j];
+      }
+      host_input_stride[i_input] = temp_stride;
+    }
+    int64_t* input_stride;
+    cudaMalloc(reinterpret_cast<void**>(&input_stride),
+               input_stride_size * sizeof(int64_t));
+    cudaMemcpy(reinterpret_cast<void*>(input_stride),
+               reinterpret_cast<void*>(host_input_stride),
+               input_stride_size * sizeof(int64_t), cudaMemcpyHostToDevice);
```
Comment (zhiqiu) [inline paddle/fluid/operators/diagonal_op.cu]: Do not use cudaMemcpy/cudaMalloc in op kernel directly. Firstly, try to use less memory as possible, i.e., do not malloc extra memory if can be avoid by refine the algorithm. Secondly, If  temporary memory is exactly needed, use temp Tensor, and call `tensor.mutable_data`.

File: paddle/fluid/operators/diagonal_op.cu [hunk 8]
```diff
@@ -0,0 +1,362 @@
...
+    }
+    cudaFree(input_stride);
+    cudaFree(output_stride);
+    free(host_input_stride);
+    free(host_output_stride);
+  }
+};
+
+template <typename T, int DX_DIM_SIZE, int DOUT_DIM_SIZE>
+__global__ void GradDiagonal(const T* dout_data, T* dx_data,
```
Comment (zhiqiu) [inline paddle/fluid/operators/diagonal_op.cu]: Most of the code in Diagonal/GradDiagonal are common and can reuse.

File: paddle/fluid/operators/diagonal_op.cc (hunk 1 above)
Comment (zhangbo9674) [reply paddle/fluid/operators/diagonal_op.cc]: Done, thanks!

File: paddle/fluid/operators/diagonal_op.cc (hunk 2 above)
Comment (zhangbo9674) [reply paddle/fluid/operators/diagonal_op.cc]: Done, thanks!

File: paddle/fluid/operators/diagonal_op.cc (hunk 3 above)
Comment (zhangbo9674) [reply paddle/fluid/operators/diagonal_op.cc]: Done, thanks!

File: paddle/fluid/operators/diagonal_op.cc (hunk 4 above)
Comment (zhangbo9674) [reply paddle/fluid/operators/diagonal_op.cc]: Done, thanks!

File: paddle/fluid/operators/diagonal_op.cc (hunk 5 above)
Comment (zhangbo9674) [reply paddle/fluid/operators/diagonal_op.cc]: Done, thanks!

File: paddle/fluid/operators/diagonal_op.cu (hunk 6 above)
Comment (zhangbo9674) [reply paddle/fluid/operators/diagonal_op.cu]: Done, thanks!

File: paddle/fluid/operators/diagonal_op.cu (hunk 7 above)
Comment (zhangbo9674) [reply paddle/fluid/operators/diagonal_op.cu]: Replace cudaMalloc/cudaMemcpy by TensorFormVector in cudaKernel and cudaGradKernel.

File: paddle/fluid/operators/diagonal_op.cu (hunk 8 above)
Comment (zhangbo9674) [reply paddle/fluid/operators/diagonal_op.cu]: Done, thanks!

File: python/paddle/tensor/__init__.py [hunk 9]
```diff
@@ -355,6 +356,8 @@
            'shape',
            'real',
            'imag',
+           'digamma',
```
Comment (TCChenlong) [inline python/paddle/tensor/__init__.py:359]: 这行要删除，和362重复

File: python/paddle/tensor/math.py [hunk 10]
```diff
@@ -1696,6 +1696,100 @@ def __check_input(input, offset, dim1, dim2):
...
+            x = paddle.rand([2,2,3],'float32')
+            print(x)
+            # Tensor(shape=[2, 2, 3], dtype=float32, place=CUDAPlace(0), stop_gradient=True,
+            #        [[[0.78954303, 0.03080254, 0.86929739],
+            #          [0.85419118, 0.66349584, 0.60896820]],
+
+            #         [[0.50427347, 0.78351408, 0.00833563],
+            #          [0.36932808, 0.83527362, 0.49408615]]])
+
+            out = paddle.diagonal(x)
```
Comment (TCChenlong) [inline python/paddle/tensor/math.py]: 需要补充以下几个示例：
- axis1 = 1的示例
- offset != 0，axis1 = 1 的示例
- offset != 0、axis1 = 1、axis2 != 1 的示例

File: python/paddle/tensor/math.py [hunk 11]
```diff
@@ -1696,6 +1696,100 @@ def __check_input(input, offset, dim1, dim2):
...
+    
+    Args:
+        x(Tensor): The input tensor x. Must be at least 2-dimensional. The input data type should be int32, int64, float32, float64.
+        offset(int, optional): Which diagonals in input tensor x will be taken. Default: 0 (main diagonals).
+        axis1(int, optional): The first axis with respect to take diagonal. Default: 0.
+        axis2(int, optional): The second axis with respect to take diagonal. Default: 1.
+        name (str, optional): Normally there is no need for user to set this property. For more information, please refer to :ref:`api_guide_Name`. Default: None.
+
+    Returns:
+        Tensor: the output data type is the same as input data type.
```
Comment (TCChenlong) [inline python/paddle/tensor/math.py]: 少了，输入 `Tensor` 在指定二维平面的局部视图 这一句

File: python/paddle/tensor/__init__.py (hunk 9 above)
Comment (zhangbo9674) [reply python/paddle/tensor/__init__.py:359]: Done, thanks!

File: python/paddle/tensor/math.py (hunk 10 above)
Comment (zhangbo9674) [reply python/paddle/tensor/math.py]: Done, thanks!

File: python/paddle/tensor/math.py (hunk 11 above)
Comment (zhangbo9674) [reply python/paddle/tensor/math.py]: Done, thanks!

File: python/paddle/tensor/math.py [hunk 12]
```diff
@@ -1696,6 +1696,116 @@ def __check_input(input, offset, dim1, dim2):
...
+            % (-(len(input_shape)), len(input_shape) - 1, axis2)
+
+        assert  axis1_ != axis2_,   \
+               "axis1 and axis2 cannot be the same axis." \
+                "But received axis1 = %d, axis2 = %d\n"%(axis1, axis2)
+
+    if in_dygraph_mode():
+        return core.ops.diagonal(x, 'offset', offset, 'axis1', axis1, 'axis2', axis2)
+
+    if not in_dygraph_mode():
```
Comment (lanxianghit) [inline python/paddle/tensor/math.py]: 这行代码是否多余？

File: python/paddle/tensor/math.py [hunk 13]
```diff
@@ -1696,6 +1696,116 @@ def __check_input(input, offset, dim1, dim2):
...
+
+            out4 = paddle.diagonal(x,0,1,2)
+            print(out4)
+            #Tensor(shape=[2, 2], dtype=float32, place=CUDAPlace(0), stop_gradient=True,
+            #       [[0.03550689, 0.28395978],
+            #        [0.39408177, 0.34103420]])
+            
+    """
+    inputs = {'Input': [x]}
+    attrs = {'offset': offset, 'axis1': axis1, 'axis2': axis2}
```
Comment (lanxianghit) [inline python/paddle/tensor/math.py]: 这两个操作为什么放在动态图调用之前？

File: python/paddle/tensor/math.py [hunk 14]
```diff
@@ -1696,6 +1696,116 @@ def __check_input(input, offset, dim1, dim2):
...
+    By default, the 2D planes formed by the first and second axis of the input tensor x.
+
+    The argument ``offset`` determines where diagonals are taken from input tensor x:
+
+    - If offset = 0, it is the main diagonal.
+    - If offset > 0, it is above the main diagonal.
+    - If offset < 0, it is below the main diagonal.
+    
+    Args:
+        x(Tensor): The input tensor x. Must be at least 2-dimensional. The input data type should be int32, int64, float32, float64.
```
Comment (lanxianghit) [inline python/paddle/tensor/math.py]: 应该考虑支持bool、fp16等更多类型

File: python/paddle/tensor/math.py (hunk 12 above)
Comment (zhangbo9674) [reply python/paddle/tensor/math.py]: 已删除，谢谢！

File: python/paddle/tensor/math.py (hunk 13 above)
Comment (zhangbo9674) [reply python/paddle/tensor/math.py]: 该段代码没有什么意义，已删除，谢谢！

File: python/paddle/tensor/math.py (hunk 14 above)
Comment (zhangbo9674) [reply python/paddle/tensor/math.py]: 已添加支持，谢谢！

File: python/paddle/tensor/math.py [hunk 15]
```diff
@@ -1696,6 +1696,114 @@ def __check_input(input, offset, dim1, dim2):
...
+    By default, the 2D planes formed by the first and second axis of the input tensor x.
+
+    The argument ``offset`` determines where diagonals are taken from input tensor x:
+
+    - If offset = 0, it is the main diagonal.
+    - If offset > 0, it is above the main diagonal.
+    - If offset < 0, it is below the main diagonal.
+    
+    Args:
+        x(Tensor): The input tensor x. Must be at least 2-dimensional. The input data type should be int32, int64, float32, float64, float16, bool.
```
Comment (lanxianghit) [inline python/paddle/tensor/math.py]:
```suggestion
        x(Tensor): The input tensor x. Must be at least 2-dimensional. The input data type should be bool, int32, int64, float16, float32, float64.
```

Comment (zhangbo9674) [reply python/paddle/tensor/math.py]: Done, thanks!

**Review by lanxianghit — DISMISSED** [review]
LGTM

**Review by TCChenlong — APPROVED** [review]
LGTM

**Review by lanxianghit — APPROVED** [review]

**Review by zhiqiu — APPROVED** [review]
LGTM

**Pull request merged**